### ISSUE TITLE
Defining tasks that depend on compiled artifacts like typechain output does not work.

### ISSUE THREAD

**Issue by KholdStare** [NomicFoundation/hardhat#2739]
There is a chicken and egg problem in defining/using tasks in hardhat:

- When `hardhat` is run (e.g. `hardhat compile`), it loads  `hardhat.config.ts`.
- Tasks need to be defined in `hardhat.config.ts` (or imported there).
- A deploy task ideally [uses the typechain types to be typesafe](https://github.com/paulrberg/solidity-template/blob/main/tasks/deploy/greeter.ts#L5-L6)
- In order for those imports to work, we need to run `hardhat compile` first. Oops! So we need to run `hardhat compile` before we can run `hardhat compile`

As an example if you look at https://github.com/paulrberg/solidity-template , the results of the typechain plugin are already checked into git: https://github.com/paulrberg/solidity-template/tree/main/src/types . They are then used in a deploy task here: https://github.com/paulrberg/solidity-template/blob/main/tasks/deploy/greeter.ts

It seems the author of above template got around that by first running compile, checking in the artifacts, and then defining the deploy task. I don't think that's a very elegant solution.

Can tasks be lazy loaded, so they are not required to be imported in `hardhat.config.ts`? I want to define tasks that use files generated by hardhat.

**Comment by KholdStare** [comment]
This issue is still relevant. The only way I was able to get around this is with dynamic import (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-4.html)

In the code it looks like:

```typescript
task("deploy:MyContract")
  .setAction(async function (taskArguments: TaskArguments, hre: HardhatRuntimeEnvironment) {
    const { deployMyContract } = await import("../../src/deployments/mycontract");

    // ...
  });
```

That way the deployment code that references the generated typescript files is not pulled in during the configuration stage. I think the configuration portion of hardhat should be broken up into "bare necessities" that are loaded just to get the CLI running, and the full blown config that's needed to actually run a deployment or compile, etc.

**Comment by fvictorio** [comment]
Hi @KholdStare, sorry for not responding before.

We are aware of this issue and agree that it's super annoying. The long-term solution is for us to just stop typechecking the project by default. This is something we've had in mind, but we didn't have an explicit issue here about it. I created [this one](https://github.com/NomicFoundation/hardhat/issues/2945) to track that.

In the meantime, I think setting `TS_NODE_TRANSPILE_ONLY=1` should help. Maybe you can add a `compile` npm script that just does `TS_NODE_TRANSPILE_ONLY=1 hardhat compile` (or use something like [`cross-env`](https://www.npmjs.com/package/cross-env), if someone in your team is on windows). Please let me know if this does (or doesn't) work for you.

**Comment by alcuadrado** [comment]
This should be fixed by v2.11.0

**Comment by marcelomorgado** [comment]
For whom that may facing this issue just use `import type { ... } from "../typechain-types"`
